holopy/scattering/theory/lens.py

```diff
@@ -0,0 +1,192 @@
+                                                        medium_index,
+                                                        illum_polarization)
+
+        fields = self._transform_integral_from_lr_to_xyz(integral_l, integral_r,
+                                                         illum_polarization)
+
+        fields *= self._compute_field_prefactor(scatterer, medium_wavevec)
+        return fields
+
+    def _compute_integral(self, positions, scatterer, medium_wavevec,
+                          medium_index, illum_polarization):
+        int_l, int_r = self._compute_integrand(positions, scatterer,
+                                               medium_wavevec, medium_index,
+                                               illum_polarization)
+        integral_l = np.sum(int_l, axis=(0,1))
+        integral_r = np.sum(int_r, axis=(0,1))
+        return integral_l, integral_r
+
+    def _compute_integrand(self, positions, scatterer, medium_wavevec,
+                           medium_index, illum_polarization):
+        krho_p, phi_p, kz_p = positions
+        pol_angle = np.arctan2(illum_polarization[1], illum_polarization[0])
+        phi_p += pol_angle.values
+        phi_p %= (2 * np.pi)
+
+        theta_shape = (self.quad_npts_theta, 1, 1)
+        th = self._theta_pts.reshape(theta_shape)
+        sinth = self._sintheta_pts.reshape(theta_shape)
+        costh = self._costheta_pts.reshape(theta_shape)
+        dth = self._theta_wts.reshape(theta_shape)
+
+        phi_shape = (1, self.quad_npts_phi, 1)
+        sinphi = self._sinphi_pts.reshape(phi_shape)
+        cosphi = self._cosphi_pts.reshape(phi_shape)
+        phi = self._phi_pts.reshape(phi_shape)
+        dphi = self._phi_wts.reshape(phi_shape)
+
+        pos_shape = (1, 1, len(kz_p))
+        krho_p = krho_p.reshape(pos_shape)
+        phi_p = phi_p.reshape(pos_shape)
+        kz_p = kz_p.reshape(pos_shape)
+
+        prefactor = self._integrand_prefactor(sinth, costh, phi, dth, dphi,
+                                              krho_p, phi_p, kz_p)
+
+        S1, S2, S3, S4 = self._calc_scattering_matrix(scatterer, medium_wavevec,
+                                                      medium_index)
+
+        integrand_l = self._integrand_prll(prefactor, cosphi, sinphi,
+                                           S1, S2, S3, S4)
+        integrand_r = self._integrand_perp(prefactor, cosphi, sinphi,
+                                           S1, S2, S3, S4)
+
+        return integrand_l, integrand_r
+
+    @classmethod
+    def _integrand_prefactor(cls, sinth, costh, phi, dth, dphi,
+                             krho_p, phi_p, kz_p):
+        if NUMEXPR_INSTALLED:
+            prefactor = ne.evaluate(cls.numexpr_integrand_prefactor1)
+            prefactor *= ne.evaluate(cls.numexpr_integrand_prefactor2)
+            prefactor *= ne.evaluate(cls.numexpr_integrand_prefactor3)
+        else:
+            prefactor = np.exp(1j * krho_p * sinth * np.cos(phi - phi_p))
```

Comment:
Is it possible to test that the two pathways produce the same result, with and without numexpr?

Reply:
alternatively you could use an eval() statement, but I don't really like that option....

Reply:
Yes, I added a few tests that do this. It's a little tricky to do the numpy part without the eval so I had to hard code the numpy expressions into the test. Basically I check that the numexpr string really do what they say the do.